ecs-init/cache/cache.go

```diff
@@ -104,28 +129,25 @@ func (d *Downloader) DownloadAgent() error {
 		return err
 	}
 
-	publishedTarballReader, err := d.getPublishedTarball()
+	tempFileName, err := d.getPublishedTarball()
 	if err != nil {
 		return err
 	}
-	defer publishedTarballReader.Close()
 
-	md5hash := md5.New()
-	tempFile, err := d.fs.TempFile(config.CacheDirectory(), "ecs-agent.tar")
+	publishedTarballReader, err := d.fs.Open(tempFileName)
 	if err != nil {
 		return err
 	}
-	log.Debugf("Temp file %s", tempFile.Name())
 	defer func() {
 		if err != nil {
-			log.Debugf("Removing temp file %s", tempFile.Name())
-			d.fs.Remove(tempFile.Name())
+			log.Debugf("Removing temp file %s", tempFileName)
```

Comment:
someone correct me if im wrong, but the `err != nil` check in the defer statement can have it's err value set to something else in L150 by `d.fs.Copy` before the defer statement actually executes. unless - is the intention to keep the tempfile around if the `d.fs.Copy` call fails? 

do we even need that `err != nil` check in the defer block?

Reply:
The intention is to remove any temp file if there's any error that fails the whole process. But I do find a mistake that I should move this defer func before its previous error check so that error from fs.Open will also result in deleting temp file.. I will update it
I think err!=nil check is needed so that we don't remove the temp file if we succeed (otherwise we will try to remove a non-existing file because it has been renamed to ecs-agent.tar)

Reply:
What do you think about doing a file existence check instead of an error check?

Your function here does the correct thing, but I'd rather not explicitly depend on scoping magic if we can avoid it.

Reply:
yeah that's probably easier to understand. I will update it